ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/OnChainPrivacyPrecompiledContract.java

```diff
@@ -0,0 +1,444 @@
+      }
+    }
+    final Hash currentBlockHash = ((BlockHeader) currentBlockHeader).getHash();
+
+    final String key = input.slice(0, 32).toBase64String();
+
+    final ReceiveResponse receiveResponse;
+    try {
+      receiveResponse = enclave.receive(key);
+    } catch (final EnclaveClientException e) {
+      LOG.debug("Can not fetch private transaction payload with key {}", key, e);
+      return Bytes.EMPTY;
+    } catch (final EnclaveServerException e) {
+      LOG.error("Enclave is responding but errored perhaps it has a misconfiguration?", e);
+      throw e;
+    } catch (final EnclaveIOException e) {
+      LOG.error("Can not communicate with enclave is it up?", e);
+      throw e;
+    }
+
+    final BytesValueRLPInput bytesValueRLPInput =
+        new BytesValueRLPInput(
+            Bytes.wrap(Base64.getDecoder().decode(receiveResponse.getPayload())), false);
+    bytesValueRLPInput.enterList();
+    final PrivateTransaction privateTransaction = PrivateTransaction.readFrom(bytesValueRLPInput);
+    final Optional<Bytes32> versionOptional;
+    if (!bytesValueRLPInput.isEndOfCurrentList()) {
+      versionOptional = Optional.of(bytesValueRLPInput.readBytes32());
+    } else {
+      versionOptional = Optional.of(Bytes32.ZERO);
+    }
+    bytesValueRLPInput.leaveList();
+
+    final WorldUpdater publicWorldState = messageFrame.getWorldState();
+
+    final Optional<Bytes> maybeGroupId = privateTransaction.getPrivacyGroupId();
+    if (maybeGroupId.isEmpty()) {
+      return Bytes.EMPTY;
+    }
+
+    final Bytes32 privacyGroupId = Bytes32.wrap(maybeGroupId.get());
+
+    LOG.trace(
+        "Processing private transaction {} in privacy group {}",
+        privateTransaction.getHash(),
+        privacyGroupId);
+
+    final PrivacyGroupHeadBlockMap privacyGroupHeadBlockMap =
+        privateStateStorage.getPrivacyGroupHeadBlockMap(currentBlockHash).orElseThrow();
+
+    final Blockchain blockchain = messageFrame.getBlockchain();
+
+    final Hash lastRootHash =
+        privateStateRootResolver.resolveLastStateRoot(privacyGroupId, currentBlockHash);
+
+    LOG.info("Last root hash: {}", lastRootHash.toHexString());
+
+    final MutableWorldState disposablePrivateState =
+        privateWorldStateArchive.getMutable(lastRootHash).get();
+
+    final WorldUpdater privateWorldStateUpdater = disposablePrivateState.updater();
+
+    maybeInjectDefaultManagementAndProxy(
+        lastRootHash, disposablePrivateState, privateWorldStateUpdater);
+
+    final PrivateTransactionProcessor.Result canExecuteResult =
+        checkCanExecute(
+            messageFrame,
+            currentBlockHeader,
+            publicWorldState,
+            privacyGroupId,
+            blockchain,
+            disposablePrivateState,
+            privateWorldStateUpdater,
+            OnChainGroupManagement.CAN_EXECUTE_METHOD_SIGNATURE);
```

Comment:
This whole logic + `canExecuteResult.getOutput().toHexString().endsWith("0")` could be encapsulated into a method isContractLocked() or something that returns a boolean. This will improve readability.

Example:

```
[...]

boolean isContractLocked = isContractLocked(messageFrame, currentBlockHeader, publicWorldState,
    privacyGroupId, blockchain, disposablePrivateState, privateWorldStateUpdater);

boolean isAddingParticipant = isCallingAddParticipantFunction(privateTransaction);

if (isAddingParticipant && !isContractLocked) {
  LOG.debug(
      "Privacy Group {} is not locked while trying to add to group with commitment {}",
      privacyGroupId.toHexString(),
      messageFrame.getTransactionHash());
  return Bytes.EMPTY;
}

if (!isAddingParticipant && isContractLocked) {
  LOG.debug(
      "Privacy Group {} is locked while trying to execute transaction with commitment {}",
      privacyGroupId.toHexString(),
      messageFrame.getTransactionHash());
  return Bytes.EMPTY;
}

[...]
```

Reply:
yep - 100% more readable!

Reply:
fixed